cursor.go

```diff
@@ -521,3 +510,110 @@ func (c *Cursor) extendLocked(response *Response) {
 
 	putResponse(response)
 }
+
+// seekCursor takes care of loading more data if needed and applying pending skips
+//
+// bufferResponse determines whether the response will be parsed into the buffer
+func (c *Cursor) seekCursor(bufferResponse bool) error {
+	if c.lastErr != nil {
+		return c.lastErr
+	}
+
+	if len(c.responses) == 0 && c.closed {
+		return errCursorClosed
+	}
+
+	// Loop over loading data, applying skips as necessary and loading more data as needed
+	// until either the cursor is closed or finished, or we have applied all outstanding
+	// skips and data is available
+	for {
+		c.applyPendingSkips(bufferResponse) // if we are buffering the responses, skip can drain from the buffer
+
+		if bufferResponse && len(c.buffer) == 0 && len(c.responses) > 0 {
+			if err := c.bufferNextResponse(); err != nil {
+				return err
+			}
+			continue // go around the loop again to re-apply pending skips
+		} else if len(c.responses) == 0 && !c.finished && !c.closed {
```

Comment:
Also I think this line should be `} else if len(c.buffer) == 0 && len(c.responses) == 0 && !c.finished && !c.closed {`